### DISCUSSION TITLE
how ipxe works in 2 DHCP server environment

### DISCUSSION THREAD

**Discussion by ydp** [ipxe/ipxe#1012] [Q&A] [answered]
Hi, I have problem to understand how ipxe works in 2 DHCP servers environment.

Previously, I used Openstack Ironic to install OS, my experiments showed that I cannot run dhcp server in network that has a existing DHCP server, if I do, the ipxe booting will randomly fail like this:
![image](https://github.com/ipxe/ipxe/assets/1532805/e7496a6e-5f42-487d-bb91-0553e21dd2b2)

Later, I switched to tinkerbell, it uses boots to provide DHCP service and claim that it can run in network that has existing DHCP server, then my experiments show it did works! and I don't even have to add ignore in existing DHCP server, I wonder how this works!

The difference is that I found that 
1. the Ironic I used use a pretty early ipxe ,  tinkerbell boots uses a much newer version, the diff is in https://github.com/ipxe/ipxe/compare/fbbdc3926...54fcb7c29
2. Ironic uses the original undionly.kpxe downloaded from ipxe.org website,  tinkerbell built its own binary with embed script https://github.com/tinkerbell/ipxedust/blob/main/binary/script/embed.ipxe 

however, even I went through those code, I still cannot figure out why it does not work in Ironic and work in tinkerbell boot.
I could observe that in the existing DHCP server without PXE feature, that it also received the DHCP discovery request and provided an offer, but it never received a DHCP request. I am not be able to understand the full meaning of this function https://github.com/ipxe/ipxe/blob/master/src/net/udp/dhcp.c#L356 .

Could anyone help explain what happens when the ipxe receives different DHCP offer from different DHCP servers?  I did not find any useful doc about this so far. Thank you in advance!

**Comment by NiKiZe** [comment]
Using 2 DHCP servers in the same subnet is in violation of specification. However if they communicate with each other, or provide separate ranges in the subnet, then it is an acceptable high availability solution, if they otherwise return same settings.

You should follow the url in the error message to understand the issue.

What I think you actually want is ProxyDHCP, which only gives boot options, see https://ipxe.org/appnote/proxydhcp

Still note that if you are using this your main DHCP server should not provide next-server or filename.

**Reply by ydp** [reply]
Hi @NiKiZe Thank you for the kind reply! Yes, proxyDHCP is a valid and better solution, but I still cannot understand how the responses from different DHCP server are handled.

In my environment, there is already a DHCP server running, as the main DHCP server, which is only capable of handling out IPs, no PXE function.  Say DHCP1.
Then I started another DHCP server with both IP allocation and PXE booting. Say DHCP2. 
The 2 DHCP servers do have overlapped IP range, because I have no privileges to modify DHCP1. (I know you mean I should not let DHCP2 to handle out IP, but I want to control the IP allocation of my machines)
In this case, if my machine starts to boot up, then will the boot fail if it receives the DHCP1 offer first, or it will just ignore the offer from DHCP1 due to no PXE info and waiting for response from DHCP2?

**Reply by NiKiZe** [reply, marked as answer]
Your scenario is not valid, therefore the result is undefined.
By having 2 DHCP servers on the same network you might see undefined behaviour in every OS. Simply don't do this.

**Reply by ydp** [reply]
Okay, thank you @NiKiZe , then there is no deterministic answer to my question. I will not do it in production, just curious why it suddenly works in my latest experiments, maybe due to networking latency or not enough experiments or any other temporary luck...
